mayastor/src/subsys/config/pool.rs

```diff
@@ -0,0 +1,181 @@
+use std::{fmt::Display, fs, path::Path};
+
+use futures::lock::Mutex;
+use once_cell::sync::{Lazy, OnceCell};
+use serde::{Deserialize, Serialize};
+
+use crate::{
+    bdev::VerboseError,
+    core::{Cores, Reactor, Share},
+    lvs::Lvs,
+    pool::PoolsIter,
+    replica::{ReplicaIter, ShareType},
+};
+
+static CONFIG_FILE: OnceCell<String> = OnceCell::new();
+
+/// Initialise the config file location
+fn init_config_location<P>(file: P)
+where
+    P: AsRef<Path> + Display + ToString,
+{
+    CONFIG_FILE.get_or_init(|| file.to_string());
+}
+
+#[inline]
+/// Return the config file location
+fn get_config_location() -> Option<&'static String> {
+    CONFIG_FILE.get()
+}
+
+#[derive(Debug, Default, PartialEq, Serialize, Deserialize)]
+#[serde(default, deny_unknown_fields)]
+pub struct PoolConfig {
+    pub pools: Option<Vec<Pool>>,
+}
+
+impl PoolConfig {
+    /// Load pool configuration from a file
+    pub fn load<P>(file: P) -> Result<PoolConfig, serde_yaml::Error>
+    where
+        P: AsRef<Path> + Display + ToString,
+    {
+        init_config_location(&file);
+
+        let bytes = fs::read(&file).unwrap_or_default();
+
+        if bytes.is_empty() {
+            return Ok(PoolConfig::default());
+        }
+
+        serde_yaml::from_slice(&bytes)
+    }
+
+    /// Write this pool configuration to a file
+    fn write<P>(&self, file: P) -> Result<(), std::io::Error>
+    where
+        P: AsRef<Path>,
+    {
+        let config = serde_yaml::to_string(&self).map_err(|error| {
+            std::io::Error::new(
+                std::io::ErrorKind::Other,
+                format!("pool configuration serialization error: {}", error),
+            )
+        })?;
+
+        fs::write(&file, config.as_bytes())
+    }
+
+    /// Export current pool configuration
+    pub async fn export() -> Result<(), std::io::Error> {
+        static MUTEX: Lazy<Mutex<u32>> = Lazy::new(|| Mutex::new(0));
+
+        if let Some(file) = get_config_location() {
+            info!("Saving pool configuration to {}", file);
+            let _guard = MUTEX.lock().await;
```

Comment:
The mutex doesn't need to be async here right?

Reply:
Somewhat moot as the following write is blocking, but at least the calling thread is not blocked while waiting to acquire the mutex.

Reply:
we shouldn't have any waiter because the the write is blocking, so no other thread should be waiting on the lock because our reactor would be "stuck" on the write.